src/parsers.rs

```diff
@@ -115,6 +136,110 @@ impl<'a> Parser<'a> {
+                // If this is not the first number or the position following
+                // a double colon, we expect there to be a single colon.
+                self.accept_char(b':')?;
+            }
+            None => (),
+        };
+
+        match self.try(|p| p.accept_number(4, 0x10000, true)) {
+            Some(part) if !use_tail && *head_idx < 8 => {
+                // Valid u16 to be added to the address
+                head[*head_idx] = part as u16;
+                *head_idx += 1;
+                Ok(())
+            },
+            Some(part) if *tail_idx < 6 => {
+                // Valid u16 to be added to the address
+                tail[*tail_idx] = part as u16;
+                *tail_idx += 1;
+                Ok(())
+            },
+            Some(_) => {
+                // Tail or head section is too long
+                Err(())
+            }
+            None if self.lookahead_char(b'/') => {
+                // We hit the prefix character. Hopefully this
+                // is a CIDR or the user used `until_eof`.
+                Ok(())
+            },
+            None if double_colon && self.pos == self.data.len() => {
+                // The address ends with "::". E.g. 1234:: or ::
+                Ok(())
+            }
+            None => {
+                // Invalid address
+                Err(())
+            }
+        }?;
+
+        if *head_idx + *tail_idx > 8 {
+            // The head and tail indexes add up to a bad address length.
+            Err(())
+        } else if !self.lookahead_char(b':') {
+            if *head_idx < 8 && !use_tail {
+                // There was no double colon found, and the head is too short
+                return Err(());
+            }
+            Ok(())
+        } else {
+            // Continue recursing
+            self.accept_ipv6_part((head, tail), (head_idx, tail_idx), use_tail)
+        }
+    }
+
+    #[cfg(feature = "ipv6")]
+    fn accept_ipv6(&mut self) -> Result<Ipv6Address> {
+        // IPv6 addresses may a "::" to indicate a series of 16 bit
+        // sections that evaluate to 0. E.g.
+        //
+        // fe80:0000:0000:0000:0000:0000:0000:0001
+        //
+        // May be written as
+        //
+        // fe80::1
+        //
+        // As a result, we need to find the first section of colon
+        // delimited u16's before a possible "::", then the
+        // possible second section after the "::", and finally
+        // combine the second optional section to the end of the
+        // final address.
+        let (mut addr, mut tail) = ([0u16; 8], [0u16; 6]);
+        let (mut head_idx, mut tail_idx) = (0, 0);
+
+        self.accept_ipv6_part((&mut addr, &mut tail), (&mut head_idx, &mut tail_idx), false)?;
+
+        // If the tail_idx is not 0, we need to copy the tail portion
+        // (the portion following the "::") to the end of the address.
+        if tail_idx != 0 {
```

Comment:
This check isn't doing anything; it would prevent a zero-length copy.

Reply:
Correct